File: project/backend/src/main/java/dev/lueem/ai/OpenAiClient.java
package dev.lueem.ai;

import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.net.URI;
import java.util.logging.Level;
import java.util.logging.Logger;
import jakarta.enterprise.context.ApplicationScoped;
import jakarta.inject.Inject;
import jakarta.json.Json;
import jakarta.json.JsonArray;
import jakarta.json.JsonObject;
import jakarta.json.JsonReader;
import java.io.StringReader;
import org.eclipse.microprofile.config.inject.ConfigProperty;

import dev.lueem.payload.PayloadBuilder;

@ApplicationScoped
public class OpenAiClient {

    @ConfigProperty(name = "OPENAI_API_KEY")
    private String openaiKey;

    private static final Logger LOGGER = Logger.getLogger(OpenAiClient.class.getName());
    private final HttpClient httpClient = HttpClient.newHttpClient();

    @Inject
    private PayloadBuilder payloadBuilder;

    private JsonArray processResponse(String responseBody, String type) {
        try (JsonReader jsonReader = Json.createReader(new StringReader(responseBody))) {
            JsonObject jsonResponse = jsonReader.readObject();
            JsonArray choices = jsonResponse.getJsonArray("choices");
            if (choices == null || choices.isEmpty()) {
                throw new RuntimeException("No choices found in the response.");
            }
            JsonObject message = choices.getJsonObject(0).getJsonObject("message");
            if (message == null) {
                throw new RuntimeException("No message found in the first choice.");
            }
            String contentString = message.getString("content");

            LOGGER.info("Content received from OpenAI: " + contentString);

            try (JsonReader contentReader = Json.createReader(new StringReader(contentString))) {
                JsonObject contentJson = contentReader.readObject();
                return contentJson.getJsonArray(type.equals("receipt") ? "ReceiptList" : "ArticleList");
            }
        } catch (Exception e) {
            LOGGER.log(Level.SEVERE, "Error processing response", e);
            throw new RuntimeException("Error processing response", e);
        }
    }

    public JsonArray askQuestion(String question, String type) {
        try {
            String payload = payloadBuilder.constructPayload(question, type);
            LOGGER.info("Payload sent to OpenAI: " + payload);

            if (openaiKey == null || openaiKey.trim().isEmpty()) {
                throw new IllegalArgumentException("API KEY is not set");
            }

            HttpRequest request = HttpRequest.newBuilder()
                    .uri(URI.create("https://api.openai.com/v1/chat/completions"))
                    .header("Content-Type", "application/json")
                    .header("Authorization", "Bearer " + openaiKey)
                    .POST(HttpRequest.BodyPublishers.ofString(payload))
                    .build();

            HttpResponse<String> response = httpClient.send(request, HttpResponse.BodyHandlers.ofString());
            LOGGER.info("Response from OpenAI: " + response.body());

            return processResponse(response.body(), type);
        } catch (Exception e) {
            LOGGER.log(Level.SEVERE, "Error in askQuestion", e);
            return Json.createArrayBuilder().build(); // Return an empty JsonArray on error
        }
    }
}

File: project/backend/src/main/kotlin/dev/lueem/shared/client/OpenAiClient.kt
package dev.lueem.shared.client

import dev.lueem.shared.config.OpenAiProperties
import dev.lueem.shared.error.UpstreamOpenAiException
import jakarta.enterprise.context.ApplicationScoped
import jakarta.inject.Inject
import jakarta.json.bind.JsonbBuilder
import java.net.URI
import java.net.http.HttpClient
import java.net.http.HttpRequest
import java.net.http.HttpResponse
import java.util.logging.Level
import java.util.logging.Logger

/**
 * HTTP client for the OpenAI chat-completions API.
 *
 * Responsibilities:
 * - Build and execute HTTP requests
 * - Translate HTTP / network errors into UpstreamOpenAiException
 * - Return the raw response body string
 *
 * Does NOT contain business logic, prompt construction, or JSON parsing.
 */
@ApplicationScoped
class OpenAiClient @Inject constructor(
    private val properties: OpenAiProperties
) {
    companion object {
        private val LOGGER = Logger.getLogger(OpenAiClient::class.java.name)
        private const val OPENAI_CHAT_URL = "https://api.openai.com/v1/chat/completions"
    }

    private val resolvedApiKey: String
        get() = properties.apiKey.takeIf { it.isNotBlank() }
            ?: throw UpstreamOpenAiException("OPENAI_API_KEY is not configured")

    private val httpClient = HttpClient.newHttpClient()
    private val jsonb = JsonbBuilder.create()

    /**
     * Executes a chat-completion request and returns the message content string.
     * Throws UpstreamOpenAiException on any failure.
     */
    fun chatCompletion(
        systemPrompt: String,
        userPrompt: String,
        model: String,
        temperature: Double
    ): String {
        val requestBody = OpenAiRequest(
            model = model,
            messages = listOf(
                OpenAiMessage("system", systemPrompt),
                OpenAiMessage("user", userPrompt)
            ),
            temperature = temperature
        )
        val responseBody = executeRequest(requestBody)
        val response = jsonb.fromJson(responseBody, OpenAiResponse::class.java)
        return response.choices?.firstOrNull()?.message?.content
            ?: throw UpstreamOpenAiException("OpenAI response missing message content")
    }

    private fun executeRequest(requestBody: OpenAiRequest): String {
        return try {
            val jsonPayload = jsonb.toJson(requestBody)
            val request = HttpRequest.newBuilder()
                .uri(URI.create(OPENAI_CHAT_URL))
                .header("Content-Type", "application/json")
                .header("Authorization", "Bearer $resolvedApiKey")
                .POST(HttpRequest.BodyPublishers.ofString(jsonPayload))
                .build()

            val response = httpClient.send(request, HttpResponse.BodyHandlers.ofString())
            val status = response.statusCode()
            if (status < 200 || status >= 300) {
                throw UpstreamOpenAiException("OpenAI HTTP $status")
            }
            response.body()
        } catch (e: UpstreamOpenAiException) {
            throw e
        } catch (e: Exception) {
            LOGGER.log(Level.SEVERE, "OpenAI request failed", e)
            throw UpstreamOpenAiException("OpenAI request failed", e)
        }
    }
}
